src/vtk_calculator.py

```diff
@@ -1,53 +1,106 @@
+    parser.add_argument("--tag", "-t", dest="tag", default="MyScalar", help="""The tag for output data.
+            Default is MyScalar""")
+    parser.add_argument("--intag", "-it", dest="intag", help="""The tag for input data.
+            Required for diff mod""")      
+    parser.add_argument("--log", "-l", dest="logging", default="INFO",
+                        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"], help="""Set the log level.
             Default is INFO""")
-
+    parser.add_argument("--diff", "-d", action='store_true', help="Calculate the difference to present data.")
+    parser.add_argument("--stats", "-s", action='store_true',
+                        help="Store stats of the difference calculation as the separate file inputmesh.stats.json")
     args = parser.parse_args()
     args.out_meshname = args.out_meshname or args.in_meshname
     return args
 
+
 def main():
     args = parse_args()
-    logging.basicConfig(level = getattr(logging, args.logging))
+    logging.basicConfig(level=getattr(logging, args.logging))
 
     assert os.path.isfile(args.in_meshname), "Input mesh file not found!"
-    data = mesh_io.read_dataset(args.in_meshname)
-    logging.info("Read in {} points.".format(data.GetNumberOfPoints()))
+
+    out_meshname = args.out_meshname
+    if args.out_meshname is None:
+        out_meshname = args.in_meshname
+
+    reader = vtk.vtkGenericDataObjectReader()
+    reader.SetFileName(args.in_meshname)
+    reader.Update()
+    vtk_dataset = reader.GetOutput()
+    logging.info("Read in {} points.".format(vtk_dataset.GetNumberOfPoints()))
 
     calc = vtk.vtkArrayCalculator()
-    calc.SetInputData(data)
+    calc.SetInputData(vtk_dataset)
     calc.AddCoordinateScalarVariable("coordsX", 0)
     calc.AddCoordinateScalarVariable("coordsY", 1)
     calc.AddCoordinateScalarVariable("coordsZ", 2)
     calc.SetFunction(args.function)
-    calc.SetResultArrayName("scalars")
-    calc.Update()
+    calc.SetResultArrayName(args.tag)
+    calc.Update()    
     logging.info("Evaluated '{}' on the mesh.".format(args.function))
+    
+    if args.diff:
+        assert args.intag is not None, "Difference mode needs input array tag"
 
-    writer = vtk.vtkUnstructuredGridWriter()
-    writer.SetInputData(calc.GetOutput())
-    writer.SetFileName(args.out_meshname)
-    writer.Write()
-    logging.info("Written output to {}.".format(args.out_meshname))
+        org_val = []
+        calc_val = []
+        orgArr = calc.GetOutput().GetPointData().GetAbstractArray(args.intag)
+        calcArr = calc.GetOutput().GetPointData().GetAbstractArray(args.tag)
+        num_points = vtk_dataset.GetNumberOfPoints()
+        for i in range(num_points):
+            org_val.append(orgArr.GetTuple1(i))
+            calc_val.append(calcArr.GetTuple1(i))
+        
+        org_val = np.array(org_val)
+        calc_val = np.array(calc_val)
+        difference = np.abs(org_val - calc_val)
```

Comment:
Wouldn't it be simpler to compute this with the same calculator as above?

You could extend the function when in diff mode to `"old - ({})".format(args.funciton)`.
Then you just need to somehow provide the _old_ value, which is `arg.intag` or `arg.tag` if the former is `None`.

Reply:
`intag` is defaulted above.

```
    if args.diff:
        calc.SetFunction('abs({}-{})'.format(intag, args.function))
        logging.info("Evaluated 'abs({}-{})' on the mesh.".format(intag, args.function))
    else:
        calc.SetFunction(args.function)
        logging.info("Evaluated '{}' on the mesh.".format(args.function))
```
It calculates function itself or difference using `vtkCalculator`

Reply:
>        calc.SetFunction('abs({}-{})'.format(intag, args.function))


Please use `intag - (function)`. Your version does not work for sums like `x+y` (`intag-x+y` vs `intag-(x+y)` ) and we need to know if the error is positive or negative (so no abs).

Reply:
Thanks for reminding parenthesis. Previous implementation `eval_mesh` was using `abs` I take that as base thats why I have abs.

Reply:
Just checked again and you are right. Sorry about that.
Then you can simply use `abs(func - intag)`.

Does this also work with vectorial data?

Reply:
Yes, it supports. I will document it.  Unit vectors are `iHat` `jHat` `kHat`.